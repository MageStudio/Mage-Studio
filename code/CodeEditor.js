import React from 'react';
import 'codemirror/lib/codemirror.css';
import './theme.scss';

import ProjectTree from './ProjectTree';
import { Col, Skeleton } from 'antd';
import { connect } from 'react-redux';
import { getScripts, getScriptContent } from '../app/actions/scripts';

let CodeMirror;
const CODEMIRROR = 'react-codemirror2';
const CODEMIRROR_JAVASCRIPT = 'codemirror/mode/javascript/javascript';

class CodeEditor extends React.Component {
    state = {
        code: '',
        loaded: false
    };

    componentDidMount() {
        const { config, getScripts } = this.props;
        const { project } = config;

        getScripts(project);

        Promise.all([
            require('react-codemirror2'),
            require('codemirror/mode/javascript/javascript')
        ]).then(([{ Controlled }]) => {
            CodeMirror = Controlled;
            this.setState({ loaded: true });
        })
    }

    handleOnBeforeChange = (editor, data, code) => {
        this.setState({ code });
    };

    handleScriptSelect = (selection) => {
        const { config } = this.props;
        const { project } = config;
        const scriptName = selection[0];

        getScriptContent(project, scriptName)
            .then(({ data }) => {
                const { content } = data;
                this.setState({ code: content });
            })
    };

    render() {
        const { scripts, config } = this.props;
        const options = {
            lineNumbers: true,
            mode: 'javascript',
            theme: 'magestudio'
        };

        return (
            <div className="code-container">
                <ProjectTree
                    config={config}
                    scripts={scripts}
                    onScriptSelect={this.handleScriptSelect}
                />
                <Col
                    span={20}
                    className="code-column">
                    { this.state.loaded ?
                        <CodeMirror
                            className='code-content'
                            value={this.state.code}
                            onBeforeChange={this.handleOnBeforeChange}
                            options={options} /> :
                        <Skeleton active />
                    }
                </Col>

            </div>
        );
    }
}

const mapStateToProps = ({ config, scripts }) => ({
    config,
    scripts
});

const mapDispatchToProps = (dispatch) => ({
    getScripts: (project) => dispatch(getScripts(project))
});

export default connect(mapStateToProps, mapDispatchToProps)(CodeEditor);
