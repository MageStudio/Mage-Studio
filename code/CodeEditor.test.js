import React from 'react';
import { shallow } from 'enzyme';

import CodeEditor from './CodeEditor';

describe('CodeEditor', () => {

    it('should render fine', () => {
        const component = shallow(<CodeEditor />);
        expect(component).toMatchSnapshot();
    });
});
