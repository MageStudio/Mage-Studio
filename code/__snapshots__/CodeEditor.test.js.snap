// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeEditor should render fine 1`] = `
<div
  className="code-container"
>
  <ProjectTree
    config={
      Object {
        "project": "test",
      }
    }
    onScriptSelect={[Function]}
    scripts={
      Object {
        "list": Array [],
      }
    }
  />
  <Col
    className="code-column"
    span={20}
  >
    <Skeleton
      active={true}
      avatar={false}
      paragraph={true}
      title={true}
    />
  </Col>
</div>
`;
